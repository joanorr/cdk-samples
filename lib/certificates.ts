import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as certificatemanager from 'aws-cdk-lib/aws-certificatemanager';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as firehose from 'aws-cdk-lib/aws-kinesisfirehose';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as targets from 'aws-cdk-lib/aws-route53-targets';
import * as s3 from 'aws-cdk-lib/aws-s3';
import {Construct} from 'constructs';
import * as path from 'path';

const DOMAIN_NAME =  'webskate101.com';
const WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME = 'Webskate101CertArn';


export class Certificates extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Reference to the pre-existing hosted zone which the API will be bound to
    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName: DOMAIN_NAME,
    });

    // Create a certificate (in this region) for API Gateways
    const cert = new certificatemanager.Certificate(this, 'ApiCert', {
      domainName: DOMAIN_NAME,
      subjectAlternativeNames: [`*.${DOMAIN_NAME}`],
      validation: {
        method: certificatemanager.ValidationMethod.DNS,
        props: {
          hostedZone: hostedZone,
        }
      }
    });

    // Export the ARN of the cert
    new cdk.CfnOutput(this, 'Webskate101CertArn', {
      exportName: WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME,
      value: cert.certificateArn,
      description: 'The ARN for the webskate101.com certificate',
    });
  }
}