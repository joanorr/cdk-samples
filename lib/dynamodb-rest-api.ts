import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as certificatemanager from 'aws-cdk-lib/aws-certificatemanager';
import * as cognito from  'aws-cdk-lib/aws-cognito';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as route53_targets from 'aws-cdk-lib/aws-route53-targets';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import {Construct} from 'constructs';
import * as path from 'path';

const WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME = 'Webskate101CertArn';
const DOMAIN_NAME =  'webskate101.com';
const SUBDOMAIN_NAME = 'dynamodb-rest-api';

// To Do:
//   * GET Card ?method=(random|least_seen|...)
//   * PUT Card ?method=(correct-answer|incorrect-answer)


export class DynamoDbRestApi extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Define the DynamoDb tables
    const decksTable = new dynamodb.TableV2(this, 'Decks', {
      partitionKey: {name: 'username', type: dynamodb.AttributeType.STRING},
      sortKey: {name: 'uid', type: dynamodb.AttributeType.STRING},
    });
    const cardsTable = new dynamodb.TableV2(this, 'Cards', {
      partitionKey: {name: 'deck', type: dynamodb.AttributeType.STRING},
      sortKey: {name: 'uid', type: dynamodb.AttributeType.STRING},
    });

    const restApi = this.initRestApi();

    // Define the resources
    const decks = restApi.root.addResource('decks');
    const cards = restApi.root.addResource('cards');

    // Define the resource handlers

    const getDecksLambda = this.initLambda('GetDecks', 'get-decks', {
      decks_table_name: decksTable.tableName});
    this.addTablePermission(getDecksLambda, decksTable, ['dynamodb:Query']);
    this.registerLambdaHandler(decks, 'GET', getDecksLambda);

    const postDecksLambda = this.initLambda('PostDecks', 'post-decks', {
      decks_table_name: decksTable.tableName});
    this.addTablePermission(postDecksLambda, decksTable, ['dynamodb:PutItem']);
    this.registerLambdaHandler(decks, 'POST', postDecksLambda);

    const postCardsLambda = this.initLambda('PostCards', 'post-cards', {
      cards_table_name: cardsTable.tableName});
    this.addTablePermission(postCardsLambda, cardsTable, ['dynamodb:PutItem']);
    this.registerLambdaHandler(cards, 'POST', postCardsLambda);

    // Set up a DNS record to the API Gateway
    new route53.ARecord(this, 'ARecord', {
      zone: route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: DOMAIN_NAME,
      }),
      recordName: SUBDOMAIN_NAME,
      target: route53.RecordTarget.fromAlias(
        new route53_targets.ApiGateway(restApi)),
      ttl: cdk.Duration.seconds(30),
    });
  }

  private initRestApi(): apigateway.RestApi {
    // The id for the Cognito user pool used for authentication
    const userPoolId = ssm.StringParameter.valueForStringParameter(
      this, '/cdk-samples/user-pool-id');

    return new apigateway.RestApi(this, 'RestApi', {
      cloudWatchRole: true,
      cloudWatchRoleRemovalPolicy: cdk.RemovalPolicy.DESTROY,
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS // this is also the default
      },
      defaultMethodOptions: {
        authorizer: new apigateway.CognitoUserPoolsAuthorizer(this, 'Auth', {
          cognitoUserPools: [
            cognito.UserPool.fromUserPoolId(this, 'UserPool', userPoolId)
          ]
        }),
        authorizationType: apigateway.AuthorizationType.COGNITO
      },
      deploy: true,
      deployOptions: {
        stageName: 'dev',
        accessLogDestination: new apigateway.LogGroupLogDestination(
          new logs.LogGroup(this, "ApiGatewayAccessLogs"))
      },
      domainName: {
        certificate: certificatemanager.Certificate.fromCertificateArn(
          this, 'Certificate',
          cdk.Fn.importValue(WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME)),
        domainName: `${SUBDOMAIN_NAME}.${DOMAIN_NAME}`,
      },
      endpointConfiguration: {
        types: [ apigateway.EndpointType.REGIONAL ]
      }
    });
  }

  private initLambda(
    name: string,
    codeDir: string,
    environment: {[key: string]: string;}) {

    return new lambda.Function(this, name, {
      runtime: lambda.Runtime.PYTHON_3_12,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(
        path.join(__dirname, '..', 'lambdas', 'dynamodb-rest-api', codeDir)),
      environment,
    });
  }

  private addTablePermission(
    fn: lambda.Function,
    table: dynamodb.TableV2,
    actions: string[]) {

    fn.addToRolePolicy(new iam.PolicyStatement({
      actions,
      effect: iam.Effect.ALLOW,
      resources: [table.tableArn]
    }));
  }

  private registerLambdaHandler(
    resource: apigateway.Resource,
    method: string,
    handler: lambda.Function): void {

    resource.addMethod(method, new apigateway.LambdaIntegration(handler));
  }

}
