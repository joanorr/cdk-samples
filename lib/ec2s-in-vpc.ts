import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elb from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import * as elb_targets from 'aws-cdk-lib/aws-elasticloadbalancingv2-targets';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as route53_targets from 'aws-cdk-lib/aws-route53-targets';
import {Construct} from 'constructs';

const DOMAIN_NAME = 'webskate101.com';
const SUBDOMAIN_NAME = 'ec2s-in-vpc';


export class Ec2sInVpc extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'VPC', {
      ipAddresses: ec2.IpAddresses.cidr('10.0.0.0/16')
    });

    const alb = new elb.ApplicationLoadBalancer(this, 'ALB', {
      vpc,
      internetFacing: true
    });

    const userData = ec2.UserData.forLinux();
    userData.addCommands(
      '#!/bin/bash',
      'yum update -y',
      'yum install -y httpd',
      'systemctl start httpd',
      'systemctl enable httpd',
      'echo "<h1>Hello from $(hostname -f)</h1>" '
          + '> /var/www/html/index.html'
    );

    // Run over the private subnets in the VPC and put an EC2 instance in each
    // one. Add each instance as a target of the ELB.
    const targets = vpc.privateSubnets.map((privateSubnet, index) => {
      const instance =  new ec2.Instance(this, `instance-${index}`, {
        vpc: vpc,
        instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
        machineImage: new ec2.AmazonLinuxImage({ generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2023 }),
        vpcSubnets: {subnets: [privateSubnet]},
        userData,
      });
      instance.connections.allowFromAnyIpv4(ec2.Port.HTTP);

      // Need to explicitly allow the ELB to connect to the instance
      alb.connections.allowTo(instance, ec2.Port.HTTP);

      // Make an ELB targetr to the instance
      return new elb_targets.InstanceIdTarget(instance.instanceId);
    });

    alb.connections.allowFromAnyIpv4(ec2.Port.HTTP);
    const listener = alb.addListener('Listener', {
      port: 80,
    });
    listener.addTargets(`HTTP targets`, {
      port: 80,
      targets
    });

    // Reference to pre-existing hosted zone
    const hostedZone = route53.HostedZone.fromLookup(
      this, 'HostedZone', {
        domainName: DOMAIN_NAME
      });
      // Set up a DNS record to the API Gateway
      new route53.ARecord(this, 'ARecord', {
        zone: hostedZone,
        recordName: SUBDOMAIN_NAME,
        target: route53.RecordTarget.fromAlias(
          new route53_targets.LoadBalancerTarget(alb)),
        ttl: cdk.Duration.seconds(30),
      });
  }
}