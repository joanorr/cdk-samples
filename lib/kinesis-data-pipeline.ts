import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as certificatemanager from 'aws-cdk-lib/aws-certificatemanager';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as kinesis from 'aws-cdk-lib/aws-kinesis';
import * as firehose from 'aws-cdk-lib/aws-kinesisfirehose';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as targets from 'aws-cdk-lib/aws-route53-targets';
import * as s3 from 'aws-cdk-lib/aws-s3';
import {Construct} from 'constructs';
import * as path from 'path';

const DOMAIN_NAME =  'webskate101.com';
const SUBDOMAIN_NAME = 'kinesis-data-pipeline';
const WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME = 'Webskate101CertArn';


export class KinesisDataPipeline extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Set up the Kinesis data stream which will receive data from the REST API
    const dataStream = new kinesis.Stream(this, 'DataStream', {
      shardCount: 1,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Set up the S3 bucket to receive data from Kinesis Firehose
    const destinationBucket = new s3.Bucket(this, 'DestinationBucket', {
      autoDeleteObjects: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    // Set up Kinesis Firehose to read from the data stream and to write to the
    // S3 bucket
    const firehoseDeliveryStreamRole = new iam.Role(
      this, 'FirehoseDeliveryStreamRole', {
        assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
        inlinePolicies: {
          firehoseDeliveryStreamIamPolicy: new iam.PolicyDocument({
            statements: [
              // See example here:
              //   https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html#using-iam-s3
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  'kinesis:DescribeStream',
                  'kinesis:GetShardIterator',
                  'kinesis:GetRecords',
                  'kinesis:ListShards',
                ],
                resources: [dataStream.streamArn],
              }),
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: [
                  's3:AbortMultipartUpload',
                  's3:GetBucketLocation',
                  's3:GetObject',
                  's3:ListBucket',
                  's3:ListBucketMultipartUploads',
                  's3:PutObject',
                ],
                resources: [
                  `${destinationBucket.bucketArn}/*`,
                ],
              }),
            ]
          }),
      }
    });
    const firehoseDeliveryStream = new firehose.CfnDeliveryStream(
      this, 'FirehoseDeliveryStream', {
        deliveryStreamType: 'KinesisStreamAsSource',
        kinesisStreamSourceConfiguration: {
          kinesisStreamArn: dataStream.streamArn,
          roleArn: firehoseDeliveryStreamRole.roleArn,
        },
        s3DestinationConfiguration: {
          bucketArn: destinationBucket.bucketArn,
          bufferingHints: {
            intervalInSeconds: 0,
            sizeInMBs: 1,
          },
          roleArn: firehoseDeliveryStreamRole.roleArn,
        },
    });

    // Set up a Lambda function to handle API requests and write data to the
    // Kinesis data stream.
    const postResponseLambda = new lambda.Function(this, 'PostResponseLambda', {
      runtime: lambda.Runtime.PYTHON_3_12,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(
        path.join(__dirname, '..', 'lambdas', 'kinesis-data-pipeline')),
      environment: {'DATA_STREAM_NAME': dataStream.streamName}
    });
    dataStream.grantWrite({
      grantPrincipal: postResponseLambda.role!.grantPrincipal
    });

    // Set up the API Gateway with an endpoint for the Lambda function
    const restApiLogGroup = new logs.LogGroup(this, "ApiGatewayAccessLogs", {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    const restApi = new apigateway.RestApi(this, 'ApiGateway', {
      endpointConfiguration: {
        types: [ apigateway.EndpointType.REGIONAL ]
      },
      cloudWatchRole: true,
      cloudWatchRoleRemovalPolicy: cdk.RemovalPolicy.DESTROY,
      deploy: true,
      deployOptions: {
        stageName: 'dev',
        accessLogDestination: new apigateway.LogGroupLogDestination(restApiLogGroup)
      },
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS // this is also the default
      },
      domainName: {
        certificate: certificatemanager.Certificate.fromCertificateArn(
          this, 'Certificate',
          cdk.Fn.importValue(WEBSKATE101_CERTIFICATE_ARN_EXPORT_NAME)),
        domainName: `${SUBDOMAIN_NAME}.${DOMAIN_NAME}`,
      },
    });
    const responses = restApi.root.addResource('responses');
    responses.addMethod('POST', new apigateway.LambdaIntegration(
      postResponseLambda));

    // Set up a DNS record to the API Gateway
    new route53.ARecord(this, 'ARecord', {
      zone: route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: DOMAIN_NAME,
      }),
      recordName: SUBDOMAIN_NAME,
      target: route53.RecordTarget.fromAlias(new targets.ApiGateway(restApi)),
      ttl: cdk.Duration.seconds(30),
    });
  }
}