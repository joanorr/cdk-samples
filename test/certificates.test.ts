import * as cdk from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import * as certificates from '../lib/certificates';

const TEST_ACCOUNT = '123456789012';
const TEST_REGION = 'eu-west-1';

let app: cdk.App;
let stack: cdk.Stack;
let template: Template;

describe('The Certificates stack', () => {
  beforeEach(() => {
    app = new cdk.App();
    stack = new certificates.Certificates(app, 'CertificatesStack', {
      env: {
        account: TEST_ACCOUNT,
        region: TEST_REGION,
      },
    });
    template = Template.fromStack(stack);
  });

  test('creates one certificate', () => {
    template.resourceCountIs('AWS::CertificateManager::Certificate', 1);
  });

  test('creates a certificate for domain name webskate101.com', () => {
    template.hasResourceProperties('AWS::CertificateManager::Certificate', {
      DomainName: 'webskate101.com',
    });
  });

  test('creates a certificate for SAN *.webskate101.com', () => {
    template.hasResourceProperties('AWS::CertificateManager::Certificate', {
      DomainName: 'webskate101.com',
      SubjectAlternativeNames: ['*.webskate101.com',],
    });
  });

  test('outputs the certificate ARN', () => {
    const cert = template.findResources(
      'AWS::CertificateManager::Certificate', {
        Properties: {DomainName: 'webskate101.com'},
    });
    const certName = Object.keys(cert)[0];

    template.hasOutput('Webskate101CertArn', {
      Export: {Name: 'Webskate101CertArn',},
      Value: {Ref: certName,},
    });
  });
});