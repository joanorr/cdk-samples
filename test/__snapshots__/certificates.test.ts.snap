// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Certificates stack passes snapshot tests 1`] = `
{
  "Outputs": {
    "Webskate101CertArn": {
      "Description": "The ARN for the webskate101.com certificate",
      "Export": {
        "Name": "Webskate101CertArn",
      },
      "Value": {
        "Ref": "ApiCert744F7693",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiCert744F7693": {
      "Properties": {
        "DomainName": "webskate101.com",
        "DomainValidationOptions": [
          {
            "DomainName": "webskate101.com",
            "HostedZoneId": "DUMMY",
          },
        ],
        "SubjectAlternativeNames": [
          "*.webskate101.com",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "CertificatesStack/ApiCert",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
